include/libp2p/protocol/kademlia/impl/kad_impl.hpp

```diff
@@ -6,36 +6,101 @@
 #ifndef LIBP2P_KAD_IMPL_HPP
 #define LIBP2P_KAD_IMPL_HPP
 
-#include <libp2p/network/network.hpp>
-#include <libp2p/peer/peer_repository.hpp>
+#include <libp2p/common/logger.hpp>
+#include <libp2p/host/host.hpp>
+#include <libp2p/protocol/kademlia/impl/kad_protocol_session.hpp>
+#include <libp2p/protocol/kademlia/impl/kad_response_handler.hpp>
 #include <libp2p/protocol/kademlia/kad.hpp>
-#include <libp2p/protocol/kademlia/message_read_writer.hpp>
-#include <libp2p/protocol/kademlia/query_runner.hpp>
 #include <libp2p/protocol/kademlia/routing_table.hpp>
 
 namespace libp2p::protocol::kademlia {
 
-  class KadImpl : public Kad {
+  class KadImpl : public Kad,
+                  public KadSessionHost,
+                  public std::enable_shared_from_this<KadImpl> {
    public:
-    ~KadImpl() override = default;
+    KadImpl(Host &host, std::shared_ptr<RoutingTable> table,
```

Comment:
Can we use shared_ptr<Host> instead reference? It is more consistent with the rest of the project and more thread safe IMO
```suggestion
    KadImpl(std::shared_ptr<Host> host, std::shared_ptr<RoutingTable> table,
```